/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.smartdata.hive.util;

import lombok.Data;
import lombok.RequiredArgsConstructor;

import java.util.Arrays;
import java.util.List;

public interface Trie<K, V> {
  boolean hasPrefixValues(Key<K> key);

  boolean hasChildValues(Key<K> key);

  boolean putIfNoIntersectingLocks(Key<K> key, V value);

  boolean remove(Key<K> key);

  static <K, V> Trie<K, V> synchronize(Trie<K, V> trie) {
    return SynchronizedTrie.wrap(trie);
  }

  @Data
  @RequiredArgsConstructor
  class Key<K> {
    private final List<K> segments;

    @SafeVarargs
    public Key(K... segments) {
      this.segments = Arrays.asList(segments);
    }
  }

  interface Node<K, V> {
    K getKey();

    V getValue();
  }

}
